app/src/sharedTest/java/org/oppia/app/utility/DragViewAction.kt

```diff
@@ -0,0 +1,173 @@
+import org.hamcrest.Matcher
+import org.hamcrest.core.AllOf.allOf
+
+/**
+ * The number of move events to send for each drag.
+ */
+private const val DRAG_STEP_COUNT = 10
+
+/**
+ * Length of time a drag should last for, in milliseconds.
+ */
+private const val DRAG_DURATION = 1500
+
+/**
+ * Duration between the last move event and the up event, in milliseconds.
+ */
+private const val WAIT_BEFORE_SENDING_UP = 400
+
+/**
+ * A custom [ViewAction] that can be replicate the long press & drag action for espresso.
+ *
+ * Reference: https://android.googlesource.com/platform/frameworks/base/+/2ff41d4/core/tests/coretests/src/android/widget/espresso/DragAction.java
+ */
+class DragViewAction(
+  private val startCoordinatesProvider: CoordinatesProvider,
+  private val endCoordinatesProvider: CoordinatesProvider,
+  private val precisionDescriber: PrecisionDescriber
+) : ViewAction {
+
+  override fun getDescription(): String = "long press and drag"
+
+  override fun getConstraints(): Matcher<View> = allOf(isCompletelyDisplayed())
+
+  override fun perform(uiController: UiController, view: View) {
+    val startCoordinates = startCoordinatesProvider.calculateCoordinates(view)
+    val endCoordinates = endCoordinatesProvider.calculateCoordinates(view)
+    val precision = precisionDescriber.describePrecision()
+    val downEvent = MotionEvents.sendDown(uiController, startCoordinates, precision).down
+    var status: Swiper.Status
+
+    try {
+      // Factor 1.5 is needed, otherwise a long press is not safely detected.
+      val longPressTimeout = (ViewConfiguration.getLongPressTimeout() * 1.5f).toLong()
+      uiController.loopMainThreadForAtLeast(longPressTimeout)
+
+      val steps = interpolate(startCoordinates, endCoordinates)
+      val delayBetweenMovements = DRAG_DURATION / steps.size
+
+      steps.forEachIndexed { index, step ->
+        if (!MotionEvents.sendMovement(uiController, downEvent, step)) {
+          MotionEvents.sendCancel(uiController, downEvent)
+          status = Swiper.Status.FAILURE
+        }
+        val desiredTime: Long = downEvent.downTime + delayBetweenMovements * index
+        val timeUntilDesired = desiredTime - SystemClock.uptimeMillis()
+        if (timeUntilDesired > 10) {
+          // If the wait time until the next event isn't long enough, skip the wait
+          // and execute the next event.
+          uiController.loopMainThreadForAtLeast(timeUntilDesired)
+        }
+      }
+      // Wait before sending up because some drag handling logic may discard move events
+      // that has been sent immediately before the up event. e.g. HandleView.
+      uiController.loopMainThreadForAtLeast(WAIT_BEFORE_SENDING_UP.toLong())
+      status = if (!MotionEvents.sendUp(uiController, downEvent, endCoordinates)) {
+        MotionEvents.sendCancel(uiController, downEvent)
+        Swiper.Status.FAILURE
+      } else {
+        Swiper.Status.SUCCESS
+      }
+    } catch (re: RuntimeException) {
```

Comment:
Why is this needed? Can't we just let the exception be thrown or is this meant to help against unreadable errors when arbitrary exceptions are thrown?

Reply:
This is done so that we can handle the errors in espresso with a proper message otherwise we wouldn't know what happened if the tests failed specially on a emulator

Reply:
Please add a line comment here to explain that since it's a bit unusual to catch RuntimeExceptions (and generally a bad practice). Feel free to resolve this comment once that change is done.

Reply:
done